### PR TITLE
[WIP] Transaction verifier

### PR DESCRIPTION
**Pull request by dconnolly**
This is a high-level sketch of what our `TransactionVerifier` should do. Right now it's calling signature batch verifiers that we may not want to use anymore, and `(Spend|Output|JoinSplit)Verifier`s that don't exist yet, and is incomplete in all its checks and outputs, but you get the idea.

### CODE REVIEW

File: zebra-chain/src/transaction/shielded_data.rs [hunk 1]
```diff
@@ -61,6 +64,32 @@ impl ShieldedData {
...
+        value_balance: Amount,
+    ) -> Result<VerificationKey<Binding>, redjubjub::Error> {
+        let cv_old: ValueCommitment = self.spends().map(|spend| spend.cv).sum();
+        let cv_new: ValueCommitment = self.outputs().map(|output| output.cv).sum();
+        let cv_balance: ValueCommitment = ValueCommitment::new(jubjub::Fr::zero(), value_balance);
+
+        let key_bytes: [u8; 32] = (cv_old - cv_new - cv_balance).into();
+
+        VerificationKey::<Binding>::try_from(key_bytes)
+    }
```
Comment (dconnolly) [inline zebra-chain/src/transaction/shielded_data.rs]: @hdevalence do you think this should live here, or is there a better place? I don't want to overload `TransactionVerifier` or whatever with like all the subroutines like this.

File: zebra-chain/src/sapling/commitment.rs [hunk 2]
```diff
@@ -137,6 +137,21 @@ impl NoteCommitment {
 #[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
 pub struct ValueCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
```
Comment (dconnolly) [inline zebra-chain/src/sapling/commitment.rs]: All these trait impls for ValueCommitment could go in their own PR @dconnolly

File: zebra-chain/src/transaction/shielded_data.rs (hunk 1 above)
Comment (hdevalence) [reply zebra-chain/src/transaction/shielded_data.rs]: Yeah, this seems like a fine place for them to go.

File: zebra-chain/src/transaction/sighash.rs [hunk 3]
```diff
@@ -24,10 +24,15 @@ const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash
...
 bitflags::bitflags! {
+    /// The different SigHash types, as defined in https://zips.z.cash/zip-0143
     pub struct HashType: u32 {
+        ///
         const ALL = 0b0000_0001;
+        ///
         const NONE = 0b0000_0010;
+        ///
         const SINGLE = Self::ALL.bits | Self::NONE.bits;
+        ///
```
Comment (hdevalence) [inline zebra-chain/src/transaction/sighash.rs:35]: The empty doc comments prevent a linter from ever noticing that they're empty, so I think it would be preferable to avoid adding them.

File: zebra-chain/src/sapling/commitment.rs (hunk 2 above)
Comment (dconnolly) [reply zebra-chain/src/sapling/commitment.rs]: https://github.com/ZcashFoundation/zebra/pull/1105

**Comment by hdevalence** [comment]
We rebased this onto #961.

File: zebra-consensus/src/primitives/groth16.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+use std::{
+    future::Future,
+    mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use jubjub;
```
Comment (dconnolly) [inline zebra-consensus/src/primitives/groth16.rs]:
```suggestion
```

File: zebra-consensus/src/transaction.rs [hunk 5]
```diff
@@ -12,11 +12,283 @@
...
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Display, Error)]
+pub enum VerifyTransactionError {
+    /// {0} version is older than the minimum transaction version: V4
+    WrongVersion(String),
+    /// transaction transfers no money
+    NoTransfer,
```
Comment (str4d) [inline zebra-consensus/src/transaction.rs:78]: This is a misleading name, as it suggests that the consensus rules verify that zero-value coins and notes cannot be spent. The actual rules that lead to this error are only checking for the presence of inputs / spends.

File: zebra-consensus/src/transaction/check.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                    .unwrap_or(true)
+            {
+                return Ok(());
+            } else {
+                return Err(VerifyTransactionError::NoTransfer);
+            }
+        }
+        _ => return Err(VerifyTransactionError::WrongVersion(tx.to_string())),
+    }
+}
```
Comment (daira) [inline zebra-consensus/src/transaction/check.rs:93]: Where's the corresponding rule for `tx_out_count`?

image

Comment (dconnolly) [reply zebra-consensus/src/transaction/check.rs:93]: https://github.com/ZcashFoundation/zebra/blob/transaction-verifier/zebra-consensus/src/transaction/check.rs#L68

**Comment by hdevalence** [comment]
Hi all, just a heads' up that this PR will be rebased and split into the part that sets up the plumbing for the transaction verification service, and the part that adds the consensus rules, which will be split off into separate PRs.  I'm hoping to get this done in the next few hours.

File: zebra-consensus/src/transaction/check.rs (hunk 6 above)
Comment (daira) [reply zebra-consensus/src/transaction/check.rs:93]: The function you linked only checks number of transparent inputs, not transparent outputs.

**Comment by dconnolly** [comment]
Closing as this has now been supplanted by #1173 and #1174
